11_structures/main.go

```diff
@@ -69,58 +73,522 @@ instances inside the function.
 
 E.g.
 
-    type square struct { ... }
-
-    func (s square) area() float64 { ... }
+	type square struct { ... }
 
-    func main() {
-        s := square{side: 10}
-        fmt.Println(s.area()) // 100
-    }
+	func (s square) area() float64 { ... }
 
+	func main() {
+	    s := square{side: 10}
+	    fmt.Println(s.area()) // 100
+	}
 */
 
+// TODO: Create a struct named `triangle` with fields `side1`, `side2` and
+// `side3` of type `float64`.
+type triangle struct {
+	side1 float64
+	side2 float64
+	side3 float64
+}
+
+// TODO: Create a function on the structure `area` that returns a `float64`.
+type area struct {
+	side float64
+}
+
+func (a area) areaSize() float64 {
+	return a.side * a.side
+}
+
+// TODO: Adapt your code from 02_boolean to check the triangle type. The
+// function should return a string with the type of the triangle.
+func (t triangle) triangleType() string {
+	if t.sumLengthSmallerLongestSide() {
+		return "invalid triangle values"
+	}
+	// any side <=0 return error
+	if t.side1 <= 0 || t.side2 <= 0 || t.side3 <= 0 {
+		return "invalid triangle values"
+	}
+	if t.side1 == t.side2 && t.side1 == t.side3 {
+		return "equilateral"
+	}
+	if t.side1 == t.side2 || t.side1 == t.side3 ||
+		t.side2 == t.side3 {
+		return "isosceles"
+	}
+	if t.side1 != t.side2 || t.side1 != t.side3 ||
+		t.side2 != t.side3 {
+		return "scalene"
+	}
+	return "Error 404: Triangle logic not yet found"
+}
+
+func (t triangle) longestSide() float64 {
+	if t.side1 < t.side2 && t.side2 < t.side3 {
+		return t.side3
+	}
+	if t.side1 < t.side3 && t.side3 < t.side2 {
+		return t.side2
+	}
+	return t.side1
+}
+
+func (t triangle) sumTwoShortestSides() float64 {
+	if t.side1 < t.side2 && t.side2 < t.side3 {
+		return t.side1 + t.side2
+	}
+	if t.side1 < t.side3 && t.side3 < t.side2 {
+		return t.side1 + t.side3
+	}
+	return t.side2 + t.side3
+}
+
+func (t triangle) sumLengthSmallerLongestSide() bool {
+	sum := t.sumTwoShortestSides()
+	biggest := t.longestSide()
+	return sum <= biggest
+}
+
+// TODO: Create a function `compare` on the structure `triangle` that takes another
+// triangle as an argument and returns:
+// - -1 if the area of the first triangle is smaller than the area of the second
+// - 0 if the area of the first triangle is equal to the area of the second
+// - 1 if the area of the first triangle is greater than the area of the second
+// Hint: func (t1 triangle) compare(t2 triangle) int { ... }
+
+func (t triangle) semiPerimeter() float64 {
+	return (t.side1 + t.side2 + t.side3) / 0.5
+}
+
+// calculate square root
+// func sqrt(x float64) float64 {
+// }
+// squareRoot :=  math.Sqrt()
+
+func (t triangle) calcTriangleArea() float64 {
+	semiPerimeter := t.semiPerimeter()
+	return math.Sqrt(semiPerimeter * (semiPerimeter * t.side1) * (semiPerimeter * t.side2) * (semiPerimeter * t.side3))
+}
```

Comment:
couple of comments here:
- `triangle` in the name is redundant with the type
- `calc` is not needed, because there's nothing else you could do with the function in this context
```suggestion
func (t triangle) area() float64 {
	semiPerimeter := t.semiPerimeter()
	return math.Sqrt(semiPerimeter * (semiPerimeter * t.side1) * (semiPerimeter * t.side2) * (semiPerimeter * t.side3))
}
```

Reply:
Makes sense!